controller/apps.go

```diff
@@ -0,0 +1,464 @@
+	}
+
+	kubeSpaceName, err := c.getSpaceNameFromSpaceID(ctx, ctx.SpaceID)
+	if err != nil {
+		return witerrors.NewNotFoundError("osio space", ctx.SpaceID.String())
+	}
+
+	// get OpenShift space
+	space, err := kc.GetSpace(*kubeSpaceName)
+	if err != nil {
+		return witerrors.NewInternalError(ctx, err)
+	}
+	if space == nil {
+		return witerrors.NewNotFoundError("space", *kubeSpaceName)
+	}
+
+	res := &app.SimpleSpaceSingle{
+		Data: space,
+	}
+
+	return ctx.OK(res)
+}
+
+// ShowSpaceApp runs the showSpaceApp action.
+func (c *AppsController) ShowSpaceApp(ctx *app.ShowSpaceAppAppsContext) error {
+
+	kc, err := c.getKubeClient(ctx)
+	if err != nil {
+		return witerrors.NewUnauthorizedError("openshift token")
+	}
+
+	kubeSpaceName, err := c.getSpaceNameFromSpaceID(ctx, ctx.SpaceID)
+	if err != nil {
+		return witerrors.NewNotFoundError("osio space", ctx.SpaceID.String())
+	}
+
+	theapp, err := kc.GetApplication(*kubeSpaceName, ctx.AppName)
+	if err != nil {
+		return witerrors.NewInternalError(ctx, err)
+	}
+	if theapp == nil {
+		return witerrors.NewNotFoundError("application", ctx.AppName)
+	}
+
+	res := &app.SimpleApplicationSingle{
+		Data: theapp,
+	}
+
+	return ctx.OK(res)
+}
+
+// ShowSpaceAppDeployment runs the showSpaceAppDeployment action.
+func (c *AppsController) ShowSpaceAppDeployment(ctx *app.ShowSpaceAppDeploymentAppsContext) error {
+
+	kc, err := c.getKubeClient(ctx)
+	if err != nil {
+		return witerrors.NewUnauthorizedError("openshift token")
+	}
+
+	kubeSpaceName, err := c.getSpaceNameFromSpaceID(ctx, ctx.SpaceID)
+	if err != nil {
+		return witerrors.NewNotFoundError("osio space", ctx.SpaceID.String())
+	}
+
+	deploymentStats, err := kc.GetDeployment(*kubeSpaceName, ctx.AppName, ctx.DeployName)
+	if err != nil {
+		return witerrors.NewInternalError(ctx, err)
+	}
+	if deploymentStats == nil {
+		return witerrors.NewNotFoundError("deployment statistics", ctx.DeployName)
+	}
+
+	res := &app.SimpleDeploymentSingle{
+		Data: deploymentStats,
+	}
+
+	return ctx.OK(res)
+}
+
+// ShowEnvAppPods runs the showEnvAppPods action.
+func (c *AppsController) ShowEnvAppPods(ctx *app.ShowEnvAppPodsAppsContext) error {
+
+	kc, err := c.getKubeClient(ctx)
+	if err != nil {
+		return witerrors.NewUnauthorizedError("openshift token")
+	}
+
+	pods, err := kc.GetPodsInNamespace(ctx.EnvName, ctx.AppName)
+	if err != nil {
+		return witerrors.NewInternalError(ctx, err)
+	}
+	if pods == nil || len(pods) == 0 {
+		return witerrors.NewNotFoundError("pods", ctx.AppName)
+	}
+	jsonresp := "{\"pods\":" + tostring(pods) + "}\n"
+
+	return ctx.OK([]byte(jsonresp))
+}
+
+// ShowSpaceEnvironments runs the showSpaceEnvironments action.
+func (c *AppsController) ShowSpaceEnvironments(ctx *app.ShowSpaceEnvironmentsAppsContext) error {
+
+	kc, err := c.getKubeClient(ctx)
+	if err != nil {
+		return witerrors.NewUnauthorizedError("openshift token")
+	}
+
+	envs, err := kc.GetEnvironments()
+	if err != nil {
+		return witerrors.NewInternalError(ctx, err)
```

Comment:
Always `errs.Wrap` the error when passing it to `NewInternalError` and annotate it please.

Reply:
Done (in followup PR)